### PR TITLE
Add boringcrypto image

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
This adds a boringcrypto image 

* builds and pushes amd64/arm64 boringcrypto image upon `make publish-images`
	* images are pushed to `grafana/rollout-operator-boringcrypto` 
	* also build the binary in an alpine container as opposed to a Debian. This means that the binary is linked against musl instead of glibc and running alpine with glibc may come with problems
* runs unit and integration tests with boringcrypto

### CODE REVIEW

**Comment by dimitarvdimitrov** [comment]
The CI is failing the same bug that @andyasp encountered here https://github.com/grafana/rollout-operator/pull/70#issuecomment-1634544728

File: Makefile
Comment (aknuds1) [inline Makefile:29]: Should you no longer push linux/arm64?

File: Makefile
Comment (aknuds1) [inline Makefile:23]: Isn't the -boringcrypto suffix missing?
```suggestion
.PHONY: build-image-boringcrypto
```

**Review by aknuds1 — COMMENTED** [review]
See questions.

File: Makefile
Comment (aknuds1) [inline Makefile:30]: Based on the PR description, are you supposed to push _only_ AMD64 for boringcrypto?

File: Makefile
Comment (dimitarvdimitrov) [reply Makefile:23]: I removed the phony altogether. I don't think we'll have a file called `build-image-boringcrypto` anyways. And IIRC we don't need .phony unless it can collide with a file name and a target

File: Makefile
Comment (dimitarvdimitrov) [reply Makefile]: yes, but I had mistakenly taken arm64 out of the wrong build command. Should be fixed now

File: Makefile
Comment (aknuds1) [inline Makefile:42]:
```suggestion
.PHONY: integration-boringcrypto
```

File: Makefile
Comment (aknuds1) [inline Makefile:44]: How do you enable boringcrypto for this test?

**Review by aknuds1 — COMMENTED** [review]
Please see comments

File: Makefile
Comment (aknuds1) [inline Makefile:45]: Extraneous newline
```suggestion
```

**Comment by dimitarvdimitrov** [comment]
thanks for the reviews @aknuds1, I think I addressed all of the issues. Sorry for the state of this PR; I should have reviewed it on my own a few times before opening.

File: .github/workflows/ci.yaml
Comment (aknuds1) [inline .github/workflows/ci.yaml:39]: [Nit] Extraneous newline
```suggestion
```

File: .github/workflows/ci.yaml
Comment (aknuds1) [inline .github/workflows/ci.yaml:50]: [Nit] Extraneous newline
```suggestion
```

File: Makefile
Comment (aknuds1) [inline Makefile:17]: Assuming that boringcrypto requires CGO.

**Review by aknuds1 — APPROVED** [review]
LGTM, modulo a couple of newline nits.

File: Makefile
Comment (dimitarvdimitrov) [reply Makefile:17]: yes, the linked openssl is in C and needs to called via CGO

Comment (dimitarvdimitrov) [reply Makefile:17]: there are more details here https://kupczynski.info/posts/fips-golang/ (is somewhat outdated because boringcrypto now supports arm64 as well as amd64, but the rest is still valid)

**Comment by dimitarvdimitrov** [comment]
I pushed a commit to pin the go version so we can see a passing CI. After that I'll undo the pin and merge on failing CI

**Comment by dimitarvdimitrov** [comment]
I'm investigating the failing CI

**Comment by dimitarvdimitrov** [comment]
I had to make some adjustment to the PR. As of now this is the amd4 binary built from this PR using the commit where CI tests ran - 7a0770d

```
$ file /tmp/rollout-operator
/tmp/rollout-operator: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, Go BuildID=GpWoX651AkwtQPsrgVNx/nuEX87cG0CpHr_FAQAlZ/WGnquJfCp-utqUX_5FNc/eePK3vlRL3Hn65mybSiG, with debug_info, not stripped
```

and includes boringcrypto & FIPS symbols as opposed to standard library symbols

```
$ go tool nm /tmp/rollout-operator | grep FIPS
  461b80 t BORINGSSL_FIPS_abort
  45da80 t FIPS_mode_set
  45da90 t FIPS_read_counter
  401490 T _cgo_d8b1bdd8e714_Cfunc__goboringcrypto_FIPS_mode
  45da70 t _goboringcrypto_FIPS_mode
  65bd40 T crypto/internal/boring._Cfunc__goboringcrypto_FIPS_mode.abi0
 2a89888 D crypto/internal/boring._cgo_d8b1bdd8e714_Cfunc__goboringcrypto_FIPS_mode
  5eadc0 T crypto/internal/boring/sig.FIPSOnly.abi0
 2a8ae30 D crypto/tls.defaultCipherSuitesFIPS
 2a8aeb0 D crypto/tls.defaultFIPSCurvePreferences
```

 The tests pass on https://github.com/grafana/rollout-operator/pull/71/commits/7a0770dea6f54c50328075874062f9ec3a3e26b2 with the go version pinned to 1.20.5. I reverted the pin in f4ef221a1041c8f073fb0c67bd9b368bacf23e99 that's why the branch is now failing CI. The failing CI is due to a bugfix in docker which hasn't reached us yet (see https://github.com/grafana/rollout-operator/pull/70#issuecomment-1634544728)

File: Dockerfile
Comment (aknuds1) [inline Dockerfile:1]: What was the purpose of `--platform=$BUILDPLATFORM`? First time I see it. Was it unnecessary?

File: Dockerfile
Comment (aknuds1) [inline Dockerfile:12]: Are libc6-compat and gcompat necessary for boringcrypto linking to work?

Could you add a comment on why they're necessary?

File: Makefile
Comment (aknuds1) [inline Makefile:29]: So boringcrypto works on ARM64 now?

**Review by aknuds1 — COMMENTED** [review]
Please see questions, otherwise good to go I think?

File: Makefile
Comment (dimitarvdimitrov) [reply Makefile:29]: boringcrypti has support for arm64, but I couldn't initially get the build going.

After making the binary dynamically linked and fixing the runtime dependencies arm64 works just as well

File: Dockerfile
Comment (dimitarvdimitrov) [reply Dockerfile:1]: the `--platform` forced to use the host os/arch even for cross-compiling. I think I ran into problems with this when compiling into CGO because the toolchain for compiling cgo on amd64 wasn't available in the arm64 golang image. Removing the `--platform=$BUILDPLATFORM` is effectively using arm64 image to build the arm64 binary and amd64 image to build the amd64 binary.

Comment (aknuds1) [reply Dockerfile:1]: Makes sense to me, thanks for the explanation.

File: Dockerfile
Comment (aknuds1) [inline Dockerfile:15]: We don't need gcompat anymore :)
```suggestion
RUN apk add --no-cache ca-certificates
```

**Review by aknuds1 — APPROVED** [review]
LGTM, although you should remove the `gcompat` package in Alpine.

File: Dockerfile
Comment (andyasp) [inline Dockerfile:1]:
```suggestion
FROM golang:1.20-alpine3.18 AS build
```
If you are linking libraries then specifying the minor alpine version matching the later stage may help prevent things from going wrong in the future if we temporarily drift from the latest version.

File: CHANGELOG.md
Comment (andyasp) [inline CHANGELOG.md:5]:
```suggestion
* [FEATURE] Publish an additional boringcrypto image for linux/amd64,linux/arm64. #71
```
Adding the word "additional" to clarify that this doesn't affect the other image.

File: CHANGELOG.md
Comment (andyasp) [inline CHANGELOG.md:6]:
```suggestion
* [ENHANCEMENT] Update the intermediate build container for the Docker image to `golang:1.20-alpine3.18`. #66 #67 #71
```
You can tell by the 3 PRs that we've really made up our mind on this.

**Review by andyasp — APPROVED** [review]
Minor comments, nice job!

File: Makefile
Comment (andyasp) [inline Makefile:29]:
```suggestion
.PHONY: publish-standard-image
```
This was my typo

File: Dockerfile
Comment (aknuds1) [reply Dockerfile]: @andyasp good catch - I was just thinking about this fact in the meantime (my fault that we didn't specify Alpine version).

**Pull request merged**
